File: Data Structures & Algorithms/combination-target-sum-ii/submission-2.java
public class Solution {
    private static List<List<Integer>> res = new ArrayList<>();

    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        res.clear();
        Arrays.sort(candidates);
        dfs(0, new ArrayList<>(), candidates, target);
        return res;
    }

    private static void dfs(int idx, List<Integer> path, int[] candidates, int target) {
        if (target == 0) {
            res.add(new ArrayList<>(path));
            return;
        }
        for (int i = idx; i < candidates.length; i++) {
            if (i > idx && candidates[i] == candidates[i - 1]) {
                continue;
            }
            if (target - candidates[i] < 0) {
                break;
            }

            path.add(candidates[i]);
            dfs(i + 1, path, candidates, target - candidates[i]);
            path.remove(path.size() - 1);
        }
    }
}

File: Data Structures & Algorithms/combination-target-sum/submission-2.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        res = []
        sub = []
        def dfs(idx, target):
            if idx == len(nums):
                if target == 0:
                    res.append(sub.copy())
                return
            
            if nums[idx] <= target:
                sub.append(nums[idx])
                dfs(idx, target - nums[idx])
                sub.pop()
            dfs(idx+1, target)
        dfs(0, target)
        return res